pkg/identity/service.go

```diff
@@ -56,9 +67,92 @@ type (
 	NotAuthenticated struct{}
 )
 
+// Initialize loads the issuer certs from disk so it can start service CSRs to proxies
+func (svc *Service) Initialize() error {
+	credentials, err := svc.loadCredentials()
+	if err != nil {
+		return err
+	}
+	svc.updateIssuer(credentials)
+	return nil
+}
+
+func (svc *Service) updateIssuer(newIssuer tls.Issuer) {
+	svc.issuerMutex.Lock()
+	svc.issuer = &newIssuer
+	log.Debug("Issuer has been updated")
+	svc.issuerMutex.Unlock()
+}
+
+// Run reads from the issuer and error channels and reloads the issuer certs when necessary
+func (svc *Service) Run(issuerEvent <-chan struct{}, issuerError <-chan error) {
+	for {
+		select {
+		case <-issuerEvent:
+			if err := svc.Initialize(); err != nil {
+				message := fmt.Sprintf("skipping issuer update as certs could not be read from disk: %s", err)
+				log.Warn(message)
+				svc.recordEvent(v1.EventTypeWarning, eventTypeSkipped, message)
+			} else {
+				message := "updated identity issuer"
+				log.Infof(message)
+				svc.recordEvent(v1.EventTypeNormal, eventTypeUpdated, message)
+			}
+		case err := <-issuerError:
+			log.Warnf("received error from fs watcher: %s", err)
+		}
+	}
+}
+
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+func (svc *Service) credPaths() (string, string) {
+	if fileExists(svc.externalIssuerPathCrt) && fileExists(svc.externalIssuerPathKey) {
+		return svc.externalIssuerPathKey, svc.externalIssuerPathCrt
+	}
+	return svc.issuerPathKey, svc.issuerPathCrt
+}
+
+func (svc *Service) loadCredentials() (tls.Issuer, error) {
+	keyPath, crtPath := svc.credPaths() //TODO: maybe cache that ?
+	creds, err := tls.ReadPEMCreds(
+		keyPath,
+		crtPath,
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to read CA from disk: %s", err)
+	}
+
+	if err := creds.Crt.Verify(svc.trustAnchors, svc.expectedName); err != nil {
+		return nil, fmt.Errorf("failed to verify issuer credentials for '%s' with trust anchors: %s", svc.expectedName, err)
+	}
+
+	log.Infof("Loaded issuer cert: %s", creds.EncodeCertificatePEM())
```

Comment:
Maybe either set this to `Debugf` or just remove it? Currently, my logs look like this:
```
linkerd-identity-7d95b7477-rvqgp identity time="2019-10-22T21:03:07Z" level=info msg="updated identity issuer"
linkerd-identity-7d95b7477-rvqgp identity time="2019-10-22T21:05:39Z" level=info msg="Loaded issuer cert: -----BEGIN CERTIFICATE-----\nMIIBrDCCAVKgAwIBAgIRANG6lRv4YowiYhSyl/AAMjkwCgYIKoZIzj0EAwIwKTEn\nMCUGA1UEAxMeaWRlbnRpdHkubGlua2VyZC5jbHVzdGVyLmxvY2FsMB4XDTE5MTAy\nMjIxMDQ1OFoXDTE5MTAyMjIyMDQ1OFowQDEVMBMGA1UEChMMY2VydC1tYW5hZ2Vy\nMScwJQYDVQQDEx5pZGVudGl0eS5saW5rZXJkLmNsdXN0ZXIubG9jYWwwWTATBgcq\nhkjOPQIBBggqhkjOPQMBBwNCAATZZJmCsTRlJ2dXbFuZcomPxwXduLuFQir8mZM3\nJ1FNLOsZCJW8tj0AHtgy2zmYcCd5U3ul6PavQa7tUzyjahobo0QwQjAOBgNVHQ8B\nAf8EBAMCAqQwDwYDVR0TAQH/BAUwAwEB/zAfBgNVHSMEGDAWgBRTBvulIY9M/owA\nlSNIgPgiUQG30TAKBggqhkjOPQQDAgNIADBFAiEAttkcAJM2BbO8y1Lgv/MN1ueR\nD1BGHvtMvWeENTN95BsCIB7ozpWURXOxksXxa7MzcpJGgjMPzvYuz8QF+1MEqwuZ\n-----END CERTIFICATE-----\n"
linkerd-identity-7d95b7477-rvqgp identity time="2019-10-22T21:05:39Z" level=info msg="updated identity issuer"
linkerd-identity-7d95b7477-rvqgp identity time="2019-10-22T21:06:41Z" level=info msg="Loaded issuer cert: -----BEGIN CERTIFICATE-----\nMIIBqzCCAVKgAwIBAgIRAIhJ4cXXxjT3yRg1PEf6PxIwCgYIKoZIzj0EAwIwKTEn\nMCUGA1UEAxMeaWRlbnRpdHkubGlua2VyZC5jbHVzdGVyLmxvY2FsMB4XDTE5MTAy\nMjIxMDU0OFoXDTE5MTAyMjIyMDU0OFowQDEVMBMGA1UEChMMY2VydC1tYW5hZ2Vy\nMScwJQYDVQQDEx5pZGVudGl0eS5saW5rZXJkLmNsdXN0ZXIubG9jYWwwWTATBgcq\nhkjOPQIBBggqhkjOPQMBBwNCAATDonFjW90A6tiLgBX/T4m+UsQe/ey97h+SNgEp\nEzixkMKLYqlrzjmuiqyrGgdwVRCk0+9uqptFrTCbRp8g8sbAo0QwQjAOBgNVHQ8B\nAf8EBAMCAqQwDwYDVR0TAQH/BAUwAwEB/zAfBgNVHSMEGDAWgBRTBvulIY9M/owA\nlSNIgPgiUQG30TAKBggqhkjOPQQDAgNHADBEAiBQtAEPgtMjwKQacOysnGP13S01\nR18xxhXo+GaUfpN28AIgYTMvKzZ0enLxELMgTDknOntx8XTpIhac8yiAMYcvCnQ=\n-----END CERTIFICATE-----\n"
```